core/src/main/java/org/apache/iceberg/DefaultValueParser.java

```
@@ -0,0 +1,190 @@
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.io.BaseEncoding;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.DateTimeUtil;
+
+public class DefaultValueParser {
+  private DefaultValueParser() {
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public static Object parseDefaultFromJson(Type type, JsonNode defaultValue) {
+
+    if (defaultValue == null || defaultValue.isNull()) {
+      return null;
+    }
+
+    switch (type.typeId()) {
+      case BOOLEAN:
+        Preconditions.checkArgument(defaultValue.isBoolean(),
+            "Cannot parse %s to a %s value", defaultValue, type);
+        return defaultValue.booleanValue();
+      case INTEGER:
+        Preconditions.checkArgument(defaultValue.isIntegralNumber() && defaultValue.canConvertToInt(),
+            "Cannot parse %s to a %s value", defaultValue, type);
+        return defaultValue.intValue();
+      case LONG:
+        Preconditions.checkArgument(defaultValue.isIntegralNumber() && defaultValue.canConvertToLong(),
+            "Cannot parse %s to a %s value", defaultValue, type);
+        return defaultValue.longValue();
+      case FLOAT:
+        Preconditions.checkArgument(defaultValue.isNumber(),
+            "Cannot parse %s to a %s value", defaultValue, type);
+        return defaultValue.floatValue();
+      case DOUBLE:
+        Preconditions.checkArgument(defaultValue.isNumber(),
+            "Cannot parse %s to a %s value", defaultValue, type);
+        return defaultValue.doubleValue();
+      case DECIMAL:
+        Preconditions.checkArgument(defaultValue.isNumber(),
+            "Cannot parse %s to a %s value", defaultValue, type);
+        return defaultValue.decimalValue();
```

I think this also needs to validate that the decimal's scale matches the expected scale. That must always match or else it should throw an exception.

---

fixed.